lib/mix/tasks/compile/phoenix_live_view.ex

```diff
@@ -92,53 +92,320 @@ defmodule Mix.Tasks.Compile.PhoenixLiveView do
+            # missing required slot
+            {nil, _slot_attr_defs} when required ->
+              message = """
+              missing required slot \"#{slot_name}\" \
+              for component #{component(call)}\
+              """
 
-              [error(message, call.file, line)]
+              [error(message, call.file, call.line)]
 
-            _ ->
+            # missing optional slot
+            {nil, _slot_attr_defs} ->
               []
+
+            # slot with attributes
+            {slot_values, slot_attr_defs} ->
+              missing_slot_attrs =
+                for slot_value <- slot_values,
+                    {attr_name, %{required: true}} <- slot_attr_defs,
+                    {line, _, _, _} = Map.fetch!(slot_value, :inner_block),
+                    not Map.has_key?(slot_value, attr_name) do
+                  message = """
+                  missing required attribute \"#{attr_name}\" \
+                  in slot \"#{slot_name}\" \
+                  for component #{component(call)}\
+                  """
+
+                  error(message, call.file, line)
+                end
+
+              slot_attrs_errors =
+                for slot_value <- slot_values,
+                    {attr_name, {line, _column, attr_kind, attr_value}} <- slot_value,
+                    attr_def = Map.get(slot_attr_defs, attr_name, :undef),
+                    reduce: [] do
+                  errors ->
+                    # type_mismatch_slot_attr_warnings =
+                    case {attr_def, attr_kind, attr_value} do
+                      # undefined attribute
+                      {:undef, _attr_kind, _attr_value} when attr_name != :inner_block ->
+                        message = """
+                        undefined attribute \"#{attr_name}\" \
+                        in slot \"#{slot_name}\" \
+                        for component #{component(call)}\
+                        """
+
+                        [error(message, call.file, line) | errors]
+
+                      # expressions cannot be type checked
+                      {%{type: _attr_type}, :expr, _attr_value} ->
+                        errors
+
+                      # string attribute type mismatch
+                      {%{type: :string}, attr_kind, attr_value}
+                      when attr_kind != :lit or not is_binary(attr_value) ->
+                        message = """
+                        attribute \"#{attr_name}\" \
+                        in slot \"#{slot_name}\" \
+                        for component #{component(call)} \
+                        must be a :string, \
+                        got: #{inspect(attr_value)}\
+                        """
+
+                        [error(message, call.file, line) | errors]
+
+                      # atom attribute type mismatch
+                      {%{type: :atom}, attr_kind, attr_value}
+                      when attr_kind != :lit or not is_atom(attr_value) ->
```

Comment:
I wonder if we could encapsulate the type logic in a function that receives `type`, `kind`, and `value` and it returns either: `:ok` or `:error`? (or even a boolean)

This will allow you to reduce the duplication in the error message as well as the duplication between slots and attrs.

Reply:
See https://github.com/phoenixframework/phoenix_live_view/pull/2073/commits/0f75ecb8465774eea0ef6c387b17b26601ad33c6 which implements this as a predicate `valid_type?/3` 🙂